packages/transaction-controller/src/TransactionController.ts

```diff
@@ -443,6 +452,27 @@ export class TransactionController extends BaseController<
     const isEIP1559Compatible = await this.getEIP1559Compatibility();
     validateTxParams(txParams, isEIP1559Compatible);
 
+    if (origin === ORIGIN_METAMASK) {
+      // Ensure the 'from' address matches the currently selected address
+      const selectedAddress = this.getSelectedAddress();
+      if (txParams.from !== selectedAddress) {
+        throw rpcErrors.internal({
+          message: `Internally initiated transaction is using invalid account.`,
+          data: {
+            origin,
+            fromAddress: txParams.from,
+            selectedAddress,
+          },
+        });
+      }
+    } else {
+      // Check if the origin has permissions to initiate transactions from the specified address
+      const permittedAddresses = await this.getPermittedAccounts(origin);
+      if (!permittedAddresses.includes(txParams.from)) {
+        throw providerErrors.unauthorized({ data: { origin: origin ?? '' } });
+      }
+    }
+
```

Comment:
Nit: Even if it's not used again, I would do it in the private function instead in order to keep `addTransaction` context cleaner.

Reply:
Absolutely, but could we go one step further and move this to the new `validation.ts` file assuming that PR goes in first?

Reply:
Updated the branch, create a separated function for this validation and moved to `validations.ts`